File: app/bidding_app/lib/services/auth_service.dart
import '../core/config/api_config.dart';
import '../models/user_model.dart';
import '../models/api_response.dart';
import 'api_client.dart';
import 'secure_storage_service.dart';

/// Authentication Service
/// 
/// Handles user authentication operations: login, register, logout, refresh token.
class AuthService {
  final ApiClient _apiClient = ApiClient();
  final SecureStorageService _storage = SecureStorageService();

  /// Register new user
  Future<ApiResponse<AuthResponse>> register({
    required String name,
    required String email,
    required String password,
    required String role,
    String? phone,
  }) async {
    final response = await _apiClient.post<AuthResponse>(
      ApiConfig.register,
      data: {
        'name': name,
        'email': email,
        'password': password,
        'role': role,
        if (phone != null) 'phone': phone,
      },
      fromJson: (json) => AuthResponse.fromJson(json),
    );

    // Save tokens and user data on success
    if (response.success && response.data != null) {
      await _storage.saveTokens(response.data!.tokens);
      await _storage.saveUser(response.data!.user);
    }

    return response;
  }

  /// Login user
  Future<ApiResponse<AuthResponse>> login({
    required String email,
    required String password,
  }) async {
    final response = await _apiClient.post<AuthResponse>(
      ApiConfig.login,
      data: {
        'email': email,
        'password': password,
      },
      fromJson: (json) => AuthResponse.fromJson(json),
    );

    // Save tokens and user data on success
    if (response.success && response.data != null) {
      await _storage.saveTokens(response.data!.tokens);
      await _storage.saveUser(response.data!.user);
    }

    return response;
  }

  /// Logout user
  Future<void> logout() async {
    try {
      // Call logout endpoint (optional)
      await _apiClient.post(
        ApiConfig.logout,
        data: {},
        fromJson: (json) => json,
      );
    } catch (e) {
      // Ignore logout errors
    } finally {
      // Clear local storage
      await _storage.clearAll();
    }
  }

  /// Check if user is logged in
  Future<bool> isLoggedIn() async {
    return await _storage.isLoggedIn();
  }

  /// Get current user
  Future<User?> getCurrentUser() async {
    return await _storage.getUser();
  }

  /// Refresh access token
  Future<bool> refreshToken() async {
    try {
      final refreshToken = await _storage.getRefreshToken();
      if (refreshToken == null) return false;

      final response = await _apiClient.post(
        ApiConfig.refresh,
        data: {'refreshToken': refreshToken},
        fromJson: (json) => json,
      );

      if (response.success && response.data != null) {
        await _storage.saveAccessToken(response.data['tokens']['accessToken']);
        await _storage.saveRefreshToken(response.data['tokens']['refreshToken']);
        return true;
      }

      return false;
    } catch (e) {
      return false;
    }
  }
}

File: backend/src/modules/auth/auth.service.ts
import User from '../../models/user.model';
import { generateTokenPair, verifyRefreshToken, JWTPayload } from '../../config/jwt';
import { RegisterDTO, LoginDTO, AuthResponseDTO } from './auth.types';
import logger from '../../utils/logger';

/**
 * Authentication Service
 * Handles business logic for authentication
 */
export class AuthService {
  /**
   * Register a new user
   */
  async register(data: RegisterDTO): Promise<AuthResponseDTO> {
    try {
      // Check if user already exists
      const existingUser = await User.findOne({ email: data.email });
      if (existingUser) {
        throw new Error('User with this email already exists');
      }

      // Create new user (password will be hashed by pre-save hook)
      const user = await User.create({
        name: data.name,
        email: data.email,
        password: data.password,
        role: data.role,
      });

      // Generate tokens
      const payload: JWTPayload = {
        userId: user._id.toString(),
        email: user.email,
        role: user.role,
      };

      const { accessToken, refreshToken } = generateTokenPair(payload);

      // Save refresh token to user
      user.refreshToken = refreshToken;
      await user.save();

      logger.info(`User registered successfully: ${user.email}`);

      return {
        user: {
          id: user._id.toString(),
          name: user.name,
          email: user.email,
          role: user.role,
          walletBalance: user.walletBalance,
          isVerified: user.isVerified,
        },
        accessToken,
        refreshToken,
      };
    } catch (error: any) {
      logger.error('Registration error:', error);
      throw error;
    }
  }

  /**
   * Login user
   */
  async login(data: LoginDTO): Promise<AuthResponseDTO> {
    try {
      // Find user and include password field
      const user = await User.findOne({ email: data.email }).select('+password');

      if (!user) {
        throw new Error('Invalid email or password');
      }

      // Verify password
      const isPasswordValid = await user.comparePassword(data.password);
      if (!isPasswordValid) {
        throw new Error('Invalid email or password');
      }

      // Generate tokens
      const payload: JWTPayload = {
        userId: user._id.toString(),
        email: user.email,
        role: user.role,
      };

      const { accessToken, refreshToken } = generateTokenPair(payload);

      // Save refresh token to user
      user.refreshToken = refreshToken;
      await user.save();

      logger.info(`User logged in: ${user.email}`);

      return {
        user: {
          id: user._id.toString(),
          name: user.name,
          email: user.email,
          role: user.role,
          walletBalance: user.walletBalance,
          isVerified: user.isVerified,
        },
        accessToken,
        refreshToken,
      };
    } catch (error: any) {
      logger.error('Login error:', error);
      throw error;
    }
  }

  /**
   * Refresh access token
   */
  async refreshAccessToken(refreshToken: string): Promise<{ accessToken: string }> {
    try {
      // Verify refresh token
      const decoded = verifyRefreshToken(refreshToken);

      // Find user and verify refresh token matches
      const user = await User.findById(decoded.userId).select('+refreshToken');

      if (!user || user.refreshToken !== refreshToken) {
        throw new Error('Invalid refresh token');
      }

      // Generate new access token
      const payload: JWTPayload = {
        userId: user._id.toString(),
        email: user.email,
        role: user.role,
      };

      const { accessToken } = generateTokenPair(payload);

      logger.debug(`Access token refreshed for user: ${user.email}`);

      return { accessToken };
    } catch (error: any) {
      logger.error('Token refresh error:', error);
      throw error;
    }
  }

  /**
   * Logout user (invalidate refresh token)
   */
  async logout(userId: string): Promise<void> {
    try {
      await User.findByIdAndUpdate(userId, { refreshToken: null });
      logger.info(`User logged out: ${userId}`);
    } catch (error: any) {
      logger.error('Logout error:', error);
      throw error;
    }
  }
}

export default new AuthService();
